KMC/kmc.hpp

```diff
@@ -1,9 +1,9 @@
 #ifndef KMC_HPP
 #define KMC_HPP
 
-#include "KMC/helpers.hpp"
```

Comment:
include "KMC/*" is used for submodule purposes with Wen's code. This allowed for easier catch2 testing and to prevent cases were people may have other helper/integration/macros/etc. header files

Reply:
Okay, I'm a little confused as to how Wen is using KMC as a submodule with these relative paths. Is Wen just copying the header files into a KMC directory in his source folder? Why can't we use consistent relative paths here?

The issue is that I don't know how I would get CMake to compile the KMC library if the include paths are broken. Could we put the other header files into a KMC subfolder so that the paths point to the header files?

Reply:
Is there a way to pull this onto a different branch besides master? I will admit that I am a complete novice when it comes to CMake, so I would like to play around with integrating it with Wen's code. There probably is a way to use the suggested relative paths but it would require changing the Makefile and a couple of other src files.

The submodule is imported outside of his src directory by the way. Here is a two level deep tree of his basic code structure
.
├── AMSOS.X
├── KMC
│   ├── KMC
│   ├── LICENSE
│   ├── Makefile
│   ├── MakefileInc.mk
│   ├── MakefileInc.mk.example
│   ├── README.md
│   ├── TEST.X
│   ├── TEST.X.dSYM
│   ├── example_objs
│   ├── kmc-documentation.cfg
│   └── tests
├── LICENSE
├── LICENSE.txt
├── Makefile
├── MakefileInc.mk
├── MakefileInc.mk.example
├── PostProcess
│   └── ParseData.py
├── Protein
│   ├── ProteinBindStatus.hpp
│   ├── ProteinConfig.cpp
│   ├── ProteinConfig.d
│   ├── ProteinConfig.hpp
│   ├── ProteinConfig.o
│   ├── ProteinConfig_example.yaml
│   ├── ProteinData.hpp
│   ├── ProteinModel.md
│   ├── ProteinTest.hpp
│   └── ProteinType.hpp
├── ProteinConfig.yaml
├── README.txt
├── RunConfig.yaml
├── SRC
│   ├── CalcProteinBind.hpp
│   ├── SystemConfig.cpp
│   ├── SystemConfig.d
│   ├── SystemConfig.hpp
│   ├── SystemConfig.o
│   ├── TubuleSystem.cpp
│   ├── TubuleSystem.d
│   ├── TubuleSystem.hpp
│   ├── TubuleSystem.o
│   ├── catch.hpp
│   ├── kmc_step_funcs.cpp
│   ├── kmc_step_funcs.d
│   ├── kmc_step_funcs.hpp
│   ├── kmc_step_funcs.o
│   ├── main.cpp
│   ├── main.d
│   ├── main.o
│   ├── test_main.cpp
│   ├── test_main.d
│   └── test_main.o
├── SimToolbox
│   ├── CMakeLists.txt
│   ├── Collision
│   ├── Doxyfile
│   ├── FDPS
│   ├── Gendoc.sh
│   ├── LICENSE
│   ├── LinuxSetup.md
│   ├── MPI
│   ├── MacSetup-deprecated.md
│   ├── Protein
│   ├── README.md
│   ├── SimToolbox
│   ├── StepByStep.md
│   ├── Sylinder
│   ├── Trilinos
│   ├── Util
│   ├── cmake
│   └── do-cmake.sh
├── TEST.X
├── TS-documentation.cfg
├── Testing
│   └── Temporary
├── build
├── doc
│   ├── html
│   └── latex

As for the references, I believe the header guards take care of the multiple include statements.

Reply:
I wasn't referring to multiple include statements being the issue, it's that `kmc.hpp` expects both `helpers.hpp` and `lookup_table.hpp` to be inside a folder named `KMC` relative to itself, but `lookup_table.hpp` also expects `helpers.hpp` to be in a folder named `KMC` relative to itself, so I feel like the actual path structure is broken in that case.

Maybe the reason `#import "KMC/foo.hpp"` might work is because there might be some absolute path structure being referenced? But I would suggest that KMC internally maintain a relative path structure that is agnostic to any global project directory that might have KMC as a dependency, in order to maintain portability.

The reason I use CMake to build KMC as a separate library is so that I can link KMC to my project and reference it like `#import <kmc.hpp>`. Used this way, having multiple headers named `helpers.hpp` never becomes an issue, because each is internal to its own project and everything is properly encapsulated. We should talk more about that on Monday and see if that would work for you and Wen.